pallets/gear-builtin/src/benchmarking.rs

```diff
@@ -101,6 +122,191 @@ benchmarks! {
+
+        // message
+        let message: G1Affine = G1::rand(&mut rng).into();
+        let message: ArkScale<Vec<<Bls12_381 as Pairing>::G1Affine>> = vec![message].into();
+        let encoded_message = message.encode();
+
+        let priv_key: ScalarField = UniformRand::rand(&mut rng);
+        let generator: G2 = G2::generator();
+        let pub_key: G2Affine = generator.mul(priv_key).into();
+        let pub_key: ArkScale<Vec<<Bls12_381 as Pairing>::G2Affine>> = vec![pub_key].into();
+        let encoded_pub_key = pub_key.encode();
+
+        let miller_loop = sp_crypto_ec_utils::bls12_381::host_calls::bls12_381_multi_miller_loop(
+            encoded_message,
+            encoded_pub_key,
+        ).unwrap();
+
+        let mut _result: Result<Vec<u8>, ()> = Err(());
+    }: {
+        _result = sp_crypto_ec_utils::bls12_381::host_calls::bls12_381_final_exponentiation(miller_loop);
+    } verify {
+        assert!(_result.is_ok());
+    }
+
+    bls12_381_msm_g1 {
+        let c in 0 .. 1_000;
+
+        let count = c as usize;
+
+        let mut rng = ark_std::test_rng();
+
+        let scalars = (0..count)
+            .map(|_| <G1 as Group>::ScalarField::rand(&mut rng))
+            .collect::<Vec<_>>();
+        let ark_scalars: ArkScale<Vec<<G1 as Group>::ScalarField>> = scalars.clone().into();
+        let encoded_scalars = ark_scalars.encode();
+
+        let bases = (0..count).map(|_| G1::rand(&mut rng)).collect::<Vec<_>>();
+        let bases = G1::batch_convert_to_mul_base(&bases);
+        let ark_bases: ArkScale<Vec<G1Affine>> = bases.clone().into();
+        let encoded_bases = ark_bases.encode();
+
+        let mut _result: Result<Vec<u8>, ()> = Err(());
+    }: {
+        _result = sp_crypto_ec_utils::bls12_381::host_calls::bls12_381_msm_g1(encoded_bases, encoded_scalars);
+    } verify {
+        let naive = naive_var_base_msm::<G1>(bases.as_slice(), scalars.as_slice());
+        let encoded = _result.unwrap();
+        let fast = ArkScaleProjective::<G1>::decode(&mut &encoded[..]).unwrap();
+        assert_eq!(naive, fast.0);
+    }
+
+    bls12_381_msm_g2 {
+        let c in 0 .. 1_000;
+
+        let count = c as usize;
+
+        let mut rng = ark_std::test_rng();
+
+        let scalars = (0..count)
+            .map(|_| <G2 as Group>::ScalarField::rand(&mut rng))
+            .collect::<Vec<_>>();
+        let ark_scalars: ArkScale<Vec<<G2 as Group>::ScalarField>> = scalars.clone().into();
+        let encoded_scalars = ark_scalars.encode();
+
+        let bases = (0..count).map(|_| G2::rand(&mut rng)).collect::<Vec<_>>();
+        let bases = G2::batch_convert_to_mul_base(&bases);
+        let ark_bases: ArkScale<Vec<G2Affine>> = bases.clone().into();
+        let encoded_bases = ark_bases.encode();
+
+        let mut _result: Result<Vec<u8>, ()> = Err(());
+    }: {
+        _result = sp_crypto_ec_utils::bls12_381::host_calls::bls12_381_msm_g2(encoded_bases, encoded_scalars);
```

Comment:
isn't it much more productive to bench exactly call including decoding instead of just host calls?

Reply:
Nope, since encoded have different length depending on bls381 entities. It means it would be quite difficult to find such items so we will get encoded length 100, 101, 102, etc